src/multiplayer.h

```diff
@@ -125,6 +125,79 @@ bool multiplayerReplicationFunctions<T>::isChanged(void* data, void* prev_data_p
 
 template <> bool multiplayerReplicationFunctions<string>::isChanged(void* data, void* prev_data_ptr);
 
+namespace replication
+{
+    class ControlBlock;
+
+    /*! a replicated item - base class that may be anything. */
+    class Item
+    {
+    public:
+        struct Settings
+        {
+            float min_delay{};
+#ifdef DEBUG
+            string name{};
+#endif
+        };
+        /*! Allow fine-grained access control to parts of the internal.
+        * 
+        * Used to grant access to some classes, while still being to guarantee class invariants.
+        */
+        class Key final {
+            friend class ControlBlock;
+            constexpr Key() = default;
+        };
+
+        explicit Item(ControlBlock& controller, const Settings& settings);
+        virtual ~Item();
+
+        virtual void send(const ControlBlock&, sf::Packet&) = 0;
+        virtual void receive(const ControlBlock&, sf::Packet&) = 0;
+
+        // 
+        ControlBlock& getController(Key) const;
+
+        uint16_t getId(Key) const;
+        void setId(Key, uint16_t id);
+    protected:
+        void setDirty();
+    private:
+        ControlBlock& controller;
+        uint16_t replication_id{};
+        // There's padding here, might be useful.
+    };
+
+    class ControlBlock
+    {
+    public:
+        class Key final {
+            friend class ReplicatedBase;
+            constexpr Key() = default;
+        };
+
+        void add(Item& item, const Item::Settings& settings);
+        bool isDirty(const Item& item) const;
+        void setDirty(const Item& item);
+        size_t send(sf::Packet&, bool everything, float delta = 0.f);
+        bool handles(uint16_t net_id) const;
+        void receive(uint16_t net_id, sf::Packet&);
+    private:
+        static constexpr uint16_t controlled_flag{ 0x8000 };
+
+        bool isDirty(size_t) const;
+        void setDirty(size_t);
+        void resetDirty();
+        std::vector<size_t> dirty;
```

Comment:
Not quite actually: this is the *structure of arrays* vs *array of structures* (SoA vs AoS) approaches.

In this particular case, it's a lot about hot/cold data separation: we know the dirty check and 'tick time' are going to be updated very often, every frame.
The members list and other constants are cold - they're only changed when members are getting registered (which is "rare" in comparisons - only happens on object construction).

So with this approach, the cache accesses are actually better: with the dirty vector packed as tightly as 1-bit per member, a single cache line holds the status for up to 512 members to be checked (assuming x64).

If all arrays were merged in a single struct, this would drop down to only ~2.5 members that can be checked for dirtiness in a single line.

Now obviously this is all part speculation, part educated guesses before any profiling comes in, and I still need this to run at scale, but the preliminary results I have are encouraging.

Reply:
When I made the comment it wasn't clear to me that the dirty is bitflags. (No idea if std::vector<bool> specialization could be used instead, but if I remember right, that one is odd and not really well defined)

As for the rest being SoA instead of AoS, I preferrer readability over optimization. Unless you can show that it makes a real bit difference on total application speed, I highly preferrer more readable code.

Reply:
If you're referring to `vector<bool>` you're right it is generally frowned up.

I was initially tempted to pull everything into a `DynamicBitset` class, hopefully that would make things more legible.
Alternatively, I considered using a `bitset<>`, but that forces either to know the replicated fields at compile time (which we should) or set an upper bound (wasn't sure if that was acceptable).